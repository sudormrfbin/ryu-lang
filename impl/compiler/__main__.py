from compiler.repl import repl

repl()

# with open("input.lang", "r") as f:
#     inp = f.read()

# output = compile(inp)
# print(output)
