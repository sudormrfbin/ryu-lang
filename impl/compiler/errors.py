from typing import ClassVar
import dataclasses
from dataclasses import dataclass
from typing_extensions import override

from lark.tree import Meta
from lark.lexer import Token

from compiler import langtypes

LineCol = tuple[int, int]


@dataclass
class Span:
    start_line: int
    end_line: int
    start_column: int
    end_column: int
    start_pos: int
    end_pos: int

    @classmethod
    def from_meta(cls, meta: Meta) -> "Span":
        return cls(
            start_line=meta.line,
            end_line=meta.end_line,
            start_column=meta.column,
            end_column=meta.end_column,
            start_pos=meta.start_pos,
            end_pos=meta.end_pos,
        )

    @classmethod
    def from_token(cls, token: Token) -> "Span":
        assert token.line is not None
        assert token.end_line is not None
        assert token.column is not None
        assert token.end_column is not None
        assert token.start_pos is not None
        assert token.end_pos is not None

        return cls(
            start_line=token.line,
            end_line=token.end_line,
            start_column=token.column,
            end_column=token.end_column,
            start_pos=token.start_pos,
            end_pos=token.end_pos,
        )

    def coord(self) -> tuple[LineCol, LineCol]:
        start = (self.start_line, self.start_column)
        end = (self.end_line, self.end_column)
        return (start, end)


class InternalCompilerError(Exception):
    pass


@dataclass
class CompilerError(Exception):
    code: ClassVar[int]

    message: str
    span: Span

    def __post_init__(self):
        attributes = dataclasses.astuple(self)
        # Custom exceptions must always call super(), we use the __post_init__
        # mechanism provided by dataclasses to call the super constructor after
        # this class's __init__ is called.
        super().__init__(*attributes)

    @override
    def __str__(self) -> str:
        return f"{type(self).__name__}: {self.message}"


@dataclass
class OperatorSpan:
    name: str
    span: Span


@dataclass
class OperandSpan:
    type_: langtypes.Type
    span: Span


@dataclass
class InvalidOperationError(CompilerError):
    code = 1

    operator: OperatorSpan
    operands: list[OperandSpan]


@dataclass
class UnknownVariable(CompilerError):
    """
    Raised when a variable is used in an expression without declaring it prior.

    ## Example
    ```
    print(var + 5)
    ```
    """

    code = 2

    variable: str


@dataclass
class UndeclaredVariable(CompilerError):
    """
    Raised when the lvalue (the variable part) of an assignment has not
    been declared prior to assignment.

    ## Example
    ```
    var = 4
    ```

    ## Fix
    ```
    let var = 4
    ```
    """

    code = 3

    variable: str


@dataclass
class UnexpectedType(CompilerError):
    """
    Raised when an expression deviates from it's expected type.

    ## Example
    ```
    if 1 {}
    ```

    ## Fix
    ```
    if true {}
    ```
    """

    code = 4

    expected_type: langtypes.Type
    actual_type: langtypes.Type


@dataclass
class TypeMismatch(CompilerError):
    """
    Raised when an expression is expected to be of a particular type
    because of a declaration in another point in the source code.

    ## Example
    ```
    match true {
        case 1 {}
        case 2 {}
    }
    ```

    ## Fix
    ```
    match 1 {
        case 1 {}
        case 2 {}
    }
    ```
    """

    code = 5

    actual_type: langtypes.Type
    expected_type: langtypes.Type
    expected_type_span: Span


@dataclass
class DuplicatedCase(CompilerError):
    """
    Raised when a match statement has duplicated case arms.

    ## Example
    ```
    match true {
        case true {}
        case false {}
    }
    ```

    ## Fix
    ```
    match true {
        case true {}
        case false {}
    }
    ```
    """

    code = 6

    previous_case_span: Span
