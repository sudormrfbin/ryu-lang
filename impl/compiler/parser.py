from typing import Any

from compiler import ast
from compiler.lalr import Token, Tree, Lark_StandAlone, v_args, Transformer, DATA

# https://github.com/lark-parser/lark/issues/565
DATA["options"]["maybe_placeholders"] = True
_parser = Lark_StandAlone(propagate_positions=True)


def parse(source: str) -> Tree[Token]:
    return _parser.parse(source)


def listify(cls):
    @v_args(meta=True, inline=False)
    def _(self, meta, lst):
        return cls(meta, lst)

    return _


@v_args(meta=True, inline=True)
class LarkTreeToAstTransformer(Transformer[Token, Any]):
    def TRUE(self, _) -> bool:
        return True

    def FALSE(self, _) -> bool:
        return False

    def INT(self, n: str) -> int:
        return int(n)

    def STRING(self, s: str) -> str:
        return s[1:-1]  # remove quotes

    statement_list = listify(ast.StatementList)
    statement_block = listify(ast.StatementBlock)
    variable_declaration = ast.VariableDeclaration
    assignment = ast.Assignment
    index_assignment = ast.IndexAssignment
    print_stmt = ast.PrintStmt
    if_chain = ast.IfChain
    if_stmt = ast.IfStmt
    else_if_ladder = listify(ast.ElseIfLadder)
    else_if_stmt = ast.ElseIfStmt
    match_stmt = ast.MatchStmt
    case_ladder = listify(ast.CaseLadder)
    case_stmt = ast.CaseStmt
    enum_pattern = ast.EnumPattern
    enum_pattern_tuple = ast.EnumPatternTuple
    wildcard_pattern = ast.WildcardPattern
    while_stmt = ast.WhileStmt
    for_stmt = ast.ForStmt
    for_stmt_int = ast.ForStmtInt
    struct_stmt = ast.StructStmt
    struct_members = listify(ast.StructMembers)
    struct_member = ast.StructMember
    struct_access = ast.StructAccess
    struct_assignment = ast.StructAssignment
    enum_stmt = ast.EnumStmt
    enum_members = listify(ast.EnumMembers)
    enum_member_bare = ast.EnumMemberBare
    enum_member_tuple = ast.EnumMemberTuple
    function_definition = ast.FunctionDefinition
    function_params = listify(ast.FunctionParams)
    function_param = ast.FunctionParam
    type_annotation = ast.TypeAnnotation
    return_stmt = ast.ReturnStmt
    equality = ast.Equality
    logical = ast.Logical
    comparison = ast.Comparison
    term = ast.Term
    factor = ast.Factor
    unary_op = ast.UnaryOp
    indexing = ast.Indexing
    function_call = ast.FunctionCall
    function_args = listify(ast.FunctionArgs)
    struct_init_members = listify(ast.StructInitMembers)
    struct_init_member = ast.StructInitMember
    bool_literal = ast.BoolLiteral
    int_literal = ast.IntLiteral
    string_literal = ast.StringLiteral
    enum_literal_simple = ast.EnumLiteralSimple
    enum_literal_tuple = ast.EnumLiteralTuple
    variable = ast.Variable
    array_literal = ast.ArrayLiteral
    array_elements = listify(ast.ArrayElements)
    array_element = ast.ArrayElement
    array_pattern = listify(ast.ArrayPattern)
    array_pattern_element = ast.ArrayPatternElement


_transformer = LarkTreeToAstTransformer()


def parse_tree_to_ast(tree: Tree[Token]) -> ast.Ast:
    return _transformer.transform(tree)
