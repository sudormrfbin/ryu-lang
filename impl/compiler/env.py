from typing import Any

from .errors import InternalCompilerError


class RuntimeEnvironment:
    values: dict[str, Any]

    def __init__(self):
        self.values = {}

    def define(self, name: str, value: Any):
        # shadowing is allowed
        self.values[name] = value

    def get(self, name: str) -> Any:
        try:
            return self.values[name]
        except KeyError:
            raise InternalCompilerError(f"Variable {name} not defined at runtime")
