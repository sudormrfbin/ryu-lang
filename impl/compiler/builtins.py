from array import array
from ctypes import Array
import ctypes
from types import BuiltinFunctionType
from typing import TYPE_CHECKING, Any
from typing_extensions import override
from compiler import langtypes
from compiler.langvalues import BuiltinFunction


if TYPE_CHECKING:
    from compiler.env import RuntimeEnvironment


class SumFunction(BuiltinFunction):
    TYPE = langtypes.Function(
        function_name="sum",
        arguments=langtypes.Params([langtypes.INT, langtypes.INT]),
        return_type=langtypes.INT,
    )

    @override
    def call(self, args: list[Any], env: "RuntimeEnvironment") -> Any:
        a: int = args[0]
        b: int = args[1]

        return a + b

class LengthFunction(BuiltinFunction):
    TYPE = langtypes.Function(
        function_name="len",
        arguments=langtypes.Params([langtypes.Array(ty=langtypes.INT)]),
        return_type=langtypes.INT
    )

    @override
    def call(self, args: list[Any], env: "RuntimeEnvironment") -> Any:
        a: list[int] = args[0] 
        
        return len(a)
        
