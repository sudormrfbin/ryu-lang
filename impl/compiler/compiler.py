from typing import TYPE_CHECKING, Any

# error_report submodule is generated dynamically by pyo3 on the rust
# side. Hence pyright cannot detect the source file.
from error_report.error_report import (  # pyright: ignore [reportMissingModuleSource]
    report_error,
)

from compiler.env import RuntimeEnvironment, TypeEnvironment

from . import errors
from .errors import (
    InvalidOperationError,
    OperandSpan,
    UndeclaredVariable,
    UnexpectedType,
    UnknownVariable,
)
from .parser import parse, parse_tree_to_ast

if TYPE_CHECKING:
    from error_report.error_report import (  # pyright: ignore [reportMissingModuleSource]
        Mark,
        Message,
    )


def run(source: str, type_env: TypeEnvironment, runtime_env: RuntimeEnvironment) -> Any:
    try:
        return _run(source, type_env, runtime_env)
    except errors.InvalidOperationError as err:
        handle_invalid_operation(err, source)
    except errors.UnknownVariable as err:
        handle_unknown_variable(err, source)
    except errors.UndeclaredVariable as err:
        handle_undeclared_variable(err, source)
    except errors.UnexpectedType as err:
        handle_unexpected_type(err, source)
    except errors.TypeMismatch as err:
        handle_type_mismatch(err, source)
    except errors.DuplicatedCase as err:
        handle_duplicated_case(err, source)
    except errors.InexhaustiveMatch as err:
        handle_inexhaustive_match(err, source)
    except errors.CompilerError as err:
        print(err)


def _run(
    source: str, type_env: TypeEnvironment, runtime_env: RuntimeEnvironment
) -> Any:
    tree = parse(source)
    ast = parse_tree_to_ast(tree)

    ast.typecheck(type_env)

    return ast.eval(runtime_env)


def handle_inexhaustive_match(err: errors.InexhaustiveMatch, source: str):
    expected_type_msg: Message = [
        "This is of type ",
        (err.expected_type.name, err.expected_type.name),
    ]
    expected_type_label: Mark = (
        expected_type_msg,
        err.expected_type.name,
        (err.expected_type_span.start_pos, err.expected_type_span.end_pos),
    )

    remaining = ", ".join((f"`{v}`" for v in err.remaining_values))
    add_block_msg: Message = [
        f"Add case block for value {remaining}"
        if len(err.remaining_values) == 1
        else f"Add case blocks for value {remaining}"
    ]
    add_block_label: Mark = (
        add_block_msg,
        (err.span.start_pos, err.span.end_pos),
    )

    labels: list[Mark] = [expected_type_label, add_block_label]
    message: Message = [
        "Match does not cover all cases for type ",
        (err.expected_type.name, err.expected_type.name),
    ]
    report_error(
        source=source,
        start_pos=err.span.start_pos,
        message=message,
        code=err.code,
        labels=labels,
    )


def handle_duplicated_case(err: errors.DuplicatedCase, source: str):
    first_occurance_msg: Message = [
        "This case is handled first here...",
    ]
    first_occurance_label: Mark = (
        first_occurance_msg,
        "color1",
        (err.previous_case_span.start_pos, err.previous_case_span.end_pos),
    )

    second_occurance_msg: Message = [
        "...and duplicated here",
    ]
    second_occurance_label: Mark = (
        second_occurance_msg,
        "color1",
        (err.span.start_pos, err.span.end_pos),
    )

    labels: list[Mark] = [first_occurance_label, second_occurance_label]
    message: Message = [
        "Duplicated case found in match expression",
    ]
    report_error(
        source=source,
        start_pos=err.span.start_pos,
        message=message,
        code=err.code,
        labels=labels,
    )


def handle_type_mismatch(err: errors.TypeMismatch, source: str):
    expected_type_msg: Message = [
        "Since this is of type ",
        (err.expected_type.name, err.expected_type.name),
        "...",
    ]
    expected_type_label: Mark = (
        expected_type_msg,
        err.expected_type.name,
        (err.expected_type_span.start_pos, err.expected_type_span.end_pos),
    )

    actual_type_msg: Message = [
        "...expected this to be ",
        (err.expected_type.name, err.expected_type.name),
        " too, but found ",
        (err.actual_type.name, err.actual_type.name),
    ]
    actual_type_label: Mark = (
        actual_type_msg,
        err.actual_type.name,
        (err.span.start_pos, err.span.end_pos),
    )

    labels: list[Mark] = [expected_type_label, actual_type_label]
    message: Message = [
        "Expected a type of ",
        (err.expected_type.name, err.expected_type.name),
        " but found ",
        (err.actual_type.name, err.actual_type.name),
    ]
    report_error(
        source=source,
        start_pos=err.span.start_pos,
        message=message,
        code=err.code,
        labels=labels,
    )


def handle_unexpected_type(err: UnexpectedType, source: str):
    actual_type_msg: Message = [
        "This is of type ",
        (err.actual_type.name, err.actual_type.name),
    ]
    actual_type_label: Mark = (
        actual_type_msg,
        err.actual_type.name,
        (err.span.start_pos, err.span.end_pos),
    )
    labels: list[Mark] = [actual_type_label]
    message: Message = [
        "Expected a type of ",
        (err.expected_type.name, err.expected_type.name),
        " but found ",
        (err.actual_type.name, err.actual_type.name),
    ]
    report_error(
        source=source,
        start_pos=err.span.start_pos,
        message=message,
        code=err.code,
        labels=labels,
    )


def handle_unknown_variable(err: UnknownVariable, source: str):
    labels: list[Mark] = [
        (["Not defined"], err.variable, (err.span.start_pos, err.span.end_pos)),
    ]
    message: Message = [
        "Variable ",
        (err.variable, err.variable),
        " not defined in this scope",
    ]
    report_error(
        source=source,
        start_pos=err.span.start_pos,
        message=message,
        code=err.code,
        labels=labels,
    )


def handle_undeclared_variable(err: UndeclaredVariable, source: str):
    labels: list[Mark] = [
        (["Not defined"], err.variable, (err.span.start_pos, err.span.end_pos)),
    ]
    message: Message = [
        "Variable ",
        (err.variable, err.variable),
        " not defined in this scope",
    ]
    report_error(
        source=source,
        start_pos=err.span.start_pos,
        message=message,
        code=err.code,
        labels=labels,
    )


def handle_invalid_operation(err: InvalidOperationError, source: str):
    labels: list[Mark] = []
    for op in err.operands:
        msg: Message = ["This is of type ", (op.type_.name, op.type_.name)]
        labels.append(
            (
                msg,
                op.type_.name,
                (op.span.start_pos, op.span.end_pos),
            )
        )

    operator = err.operator
    message: Message
    match err.operands:
        case [OperandSpan(type_=t)]:
            message = [
                f"Invalid operation '{operator.name}' for type ",
                (t.name, t.name),
            ]
        case [OperandSpan(type_=t1), OperandSpan(type_=t2)]:
            message = [
                f"Invalid operation '{operator.name}' for types ",
                (t1.name, t1.name),
                " and ",
                (t2.name, t2.name),
            ]
        case _:
            raise errors.InternalCompilerError("Unhandled case")

    report_error(
        source=source,
        start_pos=err.span.start_pos,
        message=message,
        code=err.code,
        labels=labels,
    )
