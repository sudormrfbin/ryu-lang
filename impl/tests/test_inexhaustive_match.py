import pytest
from _pytest.capture import CaptureFixture

from compiler.compiler import run
from compiler.env import RuntimeEnvironment, TypeEnvironment
from compiler.errors import InexhaustiveMatch
from compiler.parser import parse, parse_tree_to_ast
from tests.utils import multiline_sanitize

EMPTY_ENV = RuntimeEnvironment()
EMPTY_TYPE_ENV = TypeEnvironment()


SOURCE = multiline_sanitize(
    """
match true {
    case true {
        1
    }
}
"""
)


def test_match_inexhaustive_match():
    with pytest.raises(InexhaustiveMatch) as excinfo:
        ast = parse_tree_to_ast(parse(SOURCE))
        ast.typecheck(EMPTY_TYPE_ENV)

    err = excinfo.value

    assert err.span.coord() == ((1, 1), (5, 2))
    assert err.remaining_values == {False}


def test_only_variable_output(capfd: CaptureFixture[str], snapshot: str):
    run(SOURCE, EMPTY_TYPE_ENV, EMPTY_ENV)
    _, err = capfd.readouterr()
    assert snapshot == err
