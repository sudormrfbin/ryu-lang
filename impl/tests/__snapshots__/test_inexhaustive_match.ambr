# serializer version: 1
# name: test_only_variable_output
  '''
  
  [31m[E07] Error:[0m Match does not cover all cases for type [38;5;201mBool[0m
     [38;5;246mâ•­[0m[38;5;246mâ”€[0m[38;5;246m[[0mstdin:1:1[38;5;246m][0m
     [38;5;246mâ”‚[0m
   [38;5;246m1 â”‚[0m â•­â”€â–¶ match [38;5;201mt[0m[38;5;201mr[0m[38;5;201mu[0m[38;5;201me[0m {
   [38;5;240m  â”‚[0m â”‚         [38;5;201mâ”€[0m[38;5;201mâ”€[0m[38;5;201mâ”¬[0m[38;5;201mâ”€[0m  
   [38;5;240m  â”‚[0m â”‚           [38;5;201mâ•°[0m[38;5;201mâ”€[0m[38;5;201mâ”€[0m[38;5;201mâ”€[0m This is of type [38;5;201mBool[0m
   [38;5;240m  â”†[0m â”†   
   [38;5;246m5 â”‚[0m â”œâ”€â–¶ }
   [38;5;240m  â”‚[0m â”‚      
   [38;5;240m  â”‚[0m â•°â”€â”€â”€â”€â”€â”€ Add case block for value `False`
  [38;5;246mâ”€â”€â”€â•¯[0m
  
  
  '''
# ---
